### PR TITLE
[liburing] Adds liburing package

### PR DESCRIPTION
**Pull request by mdorier**
Adds the liburing package.

### CODE REVIEW

File: var/spack/repos/builtin/packages/liburing/package.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/liburing/package.py]:
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/liburing/package.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    to setup and teardown io_uring instances, and also a simplified interface
+    for applications that don't need (or want) to deal with the full kernel
+    side implementation."""
+
+    homepage = "https://github.com/axboe/liburing"
+    url = "https://github.com/axboe/liburing/archive/refs/tags/liburing-2.3.tar.gz"
+    git = "https://github.com/axboe/liburing.git"
+
+    version("master", branch="master")
+    version("2.3", sha256="60b367dbdc6f2b0418a6e0cd203ee0049d9d629a36706fcf91dfb9428bae23c8")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/liburing/package.py:21]: Confirmed the sha256 and the branch exists.

**Review by tldahlgren — DISMISSED** [review]
Confirmed the branch and sha256.

**Comment by tldahlgren** [comment]
FYI. I am unable to build the package:

```
$ spack install liburing
..
==> Installing liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5
==> No binary for liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5 found: installing from source
==> Fetching https://github.com/axboe/liburing/archive/refs/tags/liburing-2.3.tar.gz
==> No patches needed for liburing
==> liburing: Executing phase: 'autoreconf'
==> liburing: Executing phase: 'configure'
==> liburing: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

5 errors found in build log:
     184    /usr/WS1/dahlgren/spack/lib/spack/env/gcc/gcc  -D_GNU_SOURCE -D__SA
            NE_USERSPACE_TYPES__ -I../src/include/ -include ../config-host.h -g
             -O3 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -Wstring
            op-overflow=0 -Warray-bounds=0 -DLIBURING_BUILD_TEST -o unlink.t un
            link.c helpers.o  -L../src/ -luring -lpthread
     185    /usr/WS1/dahlgren/spack/lib/spack/env/gcc/gcc  -D_GNU_SOURCE -D__SA
            NE_USERSPACE_TYPES__ -I../src/include/ -include ../config-host.h -g
             -O3 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -Wstring
            op-overflow=0 -Warray-bounds=0 -DLIBURING_BUILD_TEST -o wakeup-hang
            .t wakeup-hang.c helpers.o  -L../src/ -luring -lpthread
     186    /usr/WS1/dahlgren/spack/lib/spack/env/gcc/gcc  -D_GNU_SOURCE -D__SA
            NE_USERSPACE_TYPES__ -I../src/include/ -include ../config-host.h -g
             -O3 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -Wstring
            op-overflow=0 -Warray-bounds=0 -DLIBURING_BUILD_TEST -o xattr.t xat
            tr.c helpers.o  -L../src/ -luring -lpthread
     187    /usr/WS1/dahlgren/spack/lib/spack/env/gcc/gcc  -D_GNU_SOURCE -D__SA
            NE_USERSPACE_TYPES__ -I../src/include/ -include ../config-host.h -g
             -O3 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -Wstring
            op-overflow=0 -Warray-bounds=0 -DLIBURING_BUILD_TEST -o single-issu
            er.t single-issuer.c helpers.o  -L../src/ -luring -lpthread
     188    /usr/WS1/dahlgren/spack/lib/spack/env/gcc/gcc  -D_GNU_SOURCE -D__SA
            NE_USERSPACE_TYPES__ -I../src/include/ -include ../config-host.h -g
             -O3 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -Wstring
            op-overflow=0 -Warray-bounds=0 -DLIBURING_BUILD_TEST -o send-zeroco
            py.t send-zerocopy.c helpers.o  -L../src/ -luring -lpthread
     189    In file included from send-zerocopy.c:17:
  >> 190    /usr/include/linux/ipv6.h:19:8: error: redefinition of 'struct in6_
            pktinfo'
     191     struct in6_pktinfo {
     192            ^~~~~~~~~~~
     193    In file included from /usr/include/arpa/inet.h:22,
     194                     from send-zerocopy.c:14:
     195    /usr/include/netinet/in.h:538:8: note: originally defined here
     196     struct in6_pktinfo
     197            ^~~~~~~~~~~
     198    In file included from send-zerocopy.c:17:
  >> 199    /usr/include/linux/ipv6.h:24:8: error: redefinition of 'struct ip6_
            mtuinfo'
     200     struct ip6_mtuinfo {
     201            ^~~~~~~~~~~
     202    In file included from /usr/include/arpa/inet.h:22,
     203                     from send-zerocopy.c:14:
     204    /usr/include/netinet/in.h:545:8: note: originally defined here
     205     struct ip6_mtuinfo
     206            ^~~~~~~~~~~
     207    send-zerocopy.c: In function 'create_socketpair_ip':
  >> 208    send-zerocopy.c:241:44: error: 'SO_ZEROCOPY' undeclared (first use 
            in this function); did you mean 'SO_ERROR'?
     209       if (setsockopt(*sock_client, SOL_SOCKET, SO_ZEROCOPY, &val, size
            of(val))) {
     210                                                ^~~~~~~~~~~
     211                                                SO_ERROR
     212    send-zerocopy.c:241:44: note: each undeclared identifier is reporte
            d only once for each function it appears in
  >> 213    make[1]: *** [send-zerocopy.t] Error 1
  ...
```

**Review by tldahlgren — DISMISSED** [review]
Approving but note I am unable to build the package per previous comment.

**Comment by mdorier** [comment]
That's odd. Could you give more information about your system and compiler? (I'm not the author of liburing)

EDIT: googling a bit about the error, it looks like a mismatch between the kernel's and libc's definitions of some structure. It seems to be unrelated to liburing, showing up for other software when built in older versions of Linux in which the kernel headers and libc headers weren't in sync.

**Comment by tldahlgren** [comment]
> That's odd. Could you give more information about your system and compiler? (I'm not the author of liburing)

Hopefully the following helps:

```
$ spack debug report
* **Spack:** 0.20.0.dev0 (828221b16f178595121821b7d0992f4b94485a8c)
* **Python:** 3.7.2
* **Platform:** linux-rhel7-broadwell
* **Concretizer:** clingo

$ cat <path>/spack-build-out.txt   # So gcc@8.3.1
...
==> [2023-03-01-09:38:41.159167] '/var/tmp/dahlgren/spack-stage/spack-stage-liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5/spack-src/configure' '--prefix=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5'
prefix                        /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5
includedir                    /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5/include
libdir                        /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5/lib
libdevdir                     /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5/lib
relativelibdir                
mandir                        /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5/man
datadir                       /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/liburing-2.3-5mjeoio73a7lioewmo4lpkck2mp662z5/share
stringop_overflow             yes
array_bounds                  yes
__kernel_rwf_t                no
__kernel_timespec             no
open_how                      no
statx                         no
glibc_statx                   no
C++                           yes
has_ucontext                  yes
NVMe uring command support    no
liburing_nolibc               no
CC                            /usr/WS1/dahlgren/spack/lib/spack/env/gcc/gcc
CXX                           /usr/WS1/dahlgren/spack/lib/spack/env/gcc/g++
==> liburing: Executing phase: 'build'
==> [2023-03-01-09:38:42.150314] 'make' '-j16' 'V=1'
...
```

**Comment by mdorier** [comment]
Yeah RHEL7 probably predates the synchronization of the kernel and libc headers.

**Comment by mdorier** [comment]
Note also that io_uring is available in Linux kernel starting from version 5.1 (RHEL7 uses 3.10). I don't know if there is a way in spack to specify such a requirement. It is also not going to work in non-linux platforms.

**Comment by tldahlgren** [comment]
> Note also that io_uring is available in Linux kernel starting from version 5.1 (RHEL7 uses 3.10). I don't know if there is a way in spack to specify such a requirement. 

That's a good question in terms of a general solution.  For RHEL, it looks like any v5.* is only supported in RHEL9 so you could add conflicts:

```
    for op_sys in ["rhel7", "rhel8"]:
        conflicts("os={0}".format(op_sys), msg="Requires Linux kernel v5.1 on'")
```

> It is also not going to work in non-linux platforms.

Here again you can add ``conflicts()`` directives to flag unsupported platforms, e.g.,:

```
    for platform in ["cray", "darwin", "windows"]:
        conflicts("platform={0}".format(platform), msg="Only support on 'linux'")
```

**Comment by mdorier** [comment]
Ok I added conflicts for darwin and windows. Some cray do have io_uring. I'm not going to try tracking all the Linux and Cray OS individually to add conflict on those that don't have support.

**Comment by tldahlgren** [comment]
> Ok I added conflicts for darwin and windows. Some cray do have io_uring. I'm not going to try tracking all the Linux and Cray OS individually to add conflict on those that don't have support.

Great!  Thanks.

It might help people trying to debug failing installs to have a comment about the kernel version requirement.  It could even be added to the package docstring, perhaps, to make it visible in `spack info liburing` though I'm not sure how popular that idea would be with Spack maintainers.

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**